### ISSUE TITLE
The server should warn the app if /me/head doesn't exist

### ISSUE THREAD

**Issue by russell-taylor** [OSVR/OSVR-Core#289]
It seems like maybe it should refuse to run to make sure that people know there is a problem.

Otherwise, all apps probably just need to check once they have asked for it to see if it exists.

**Comment by godbyk** [comment]
I don't think it should refuse to run. I'm interested in using OSVR as a replacement for VRPN, for example, and don't need it to support an HMD in all use cases.

**Comment by janoc** [comment]
I agree with @godbyk  - we have also plenty of use cases around where there is no /me/head or an HMD. Please don't hardwire such assumptions - if the applications requires /me/head, it can easily check for its presence in the published tree and then decide what to do. It is not the the job of the middleware to decide this.

**Comment by JeroMiya** [comment]
I also agree. /me/head is special only when using the DisplayConfig API or RenderManager. Those APIs should fail without a /me/head (and I believe they do already), but otherwise OSVR should be agnostic.

**Comment by janoc** [comment]
Even there I would be careful about a hardwired failure. /me/head is an assumption of a simplified situation where there is a *single* user. Now imagine that you have Kinect exposed via OSVR (we have an in-house OSVR server for it) and the Kinect SDK supports up to 6 simultaneous users. Which one is /me/head? It could be completely logical to have aliases /me/head0, /me/head1, etc. in such case, and no /me/head. 

Those APIs should allow to specify which device should be used if they don't already, with the /me/head being a good fallback default for the most common use case. 

(BTW, don't fall into the trap of defining /me/head0 as the "main" user or aliasing it to /me/head - Kinect SDK will give you user id 2 just fine even if no other user is visible, for example, so you won't have user 0, but there will be user 2. Such assumptions break very easily.)

**Comment by JeroMiya** [comment]
for additional users, you would use a different base path. something other than /me/*. There isn't currently a convention set in stone for additional users, but it might be something like /player0/head, /player1/head, etc...

**Comment by janoc** [comment]
@JeroMiya Okay, that makes sense. However, it still doesn't change anything on the fact that in such case /me/head could be missing and the DisplayConfig API should use one of the /playerX/head ones instead (determined e.g. by some sort of initialization procedure in the app). It certainly shouldn't bail with a hardwired error message about /me/head. 

Also display could be easily a large screen (projector, TV, ...), not only an HMD.

**Comment by JeroMiya** [comment]
The kinect plugin itself, if it provides a mapping for /me/head and it makes sense to do so, should internally select among the tracked skeletons (using some kind of ranking - closest to the kinect?) and forward tracker reports through as /me/head from that skeleton. Separately it can also provide users 0-N which may or may not be "active". In this case, /me/head is a virtual user representing the "most likely" one. In OSVR, the current rule of thumb for occasionally tracked objects (like AR or skeletal trackers, etc...) is to always report that the device exists if it *could* send a report, but only send reports if they are available.

**Comment by janoc** [comment]
The problem is that from user's point of view, this sort of "being clever" is terrible - it rarely does what the user (application developer) wants and often just gets in the way.

Many applications that use Kinect (or a similar tech - e.g. Playstation Eye cameras in the past) identify the users by asking them to put their bodies into certain zones on the screen. Or ask them to perform a certain gesture, permitting to identify which user is which. That is not something envisageable to be provided by OSVR as it is very application-specific. On the other hand, in that context the /me/head pre-defined by a simplistic heuristic would be pointless and ignored in the best case but frustrating and getting in the way in the worst case (e.g. if the DisplayConfig API would insist on using the hard-wired and incorrect /me/head alias or causing the application to unconditionally abort if it is missing). 

Nobody is asking for OSVR to explicitly support multi-user use cases out of the box. However, it shouldn't make them unnecessarily difficult to implement because of an arbitrary limitation getting in the way. 

This is unfortunately a rather pervasive issue in OSVR - trying to force the developers to design their applications in certain ways which may not be applicable to their use cases, requiring kludges and workarounds. 

For example, the API still assumes that the OSVR server is collocated with the application and doesn't permit to specify the host:port combination when creating the context. Which is actually an extremely common situation from the VRPN world - the server providing trackers and input hw is usually not the same machine as the one running the application. Sometimes it is even outright impossible to collocate them - e.g. when using GearVR with an external tracking system as a wireless HMD for a mobile user. It can be worked around by using an env variable, but that's hardly ideal. There is actually an open pull request for this issue which wasn't merged yet.

**Comment by russell-taylor** [comment]
@janoc As one of the people building the OSVR plane while trying to get it off the ground, I can assure you that what may appear as attempts to coerce developers to fit their applications into a mold are in fact the result of having too many balls in the air and not enough hands -- not even enough hands to keep up with all of the excellent pull requests that are coming our way.

Our overall design goal is "To make difficult things easy without making easy things difficult or making anything impossible."  Of course, sometimes these goals are in conflict and sometimes they are not perfectly met right away.  That's what version 0.6 means.

I appreciate your advice on how to make things more general as we move towards version 1.0.  Do know that there is a horde of end users saying "I just want it to work out of the box without having to specify all of this stuff" trying to make common case easy.  Coming up with a design that meets both sets of needs is one of the juiciest design challenges there is.  Taming complexity without cutting off flexibility is one of the fundamental problems of any library and one we are striving to attain.

**Comment by janoc** [comment]
Hello Russell, 

First I should probably apologize for the tone of my previous reply to Jeremy - it likely came across more harsh than it was meant to.

However, I stand behind the point about some of the design decisions - e.g. for the one about specifying the hostname to connect to I was explicitly told it is essentially a policy decision to not have any client-side configuration, period. The discussion is here on the issue tracker, if you want to see it. The unfortunate consequence is that I have to maintain an in-house fork because we need to be able to specify a remote server (plus a few other things, such as the tree access from the client). 

I do understand your issues regarding the manpower and shooting for the most common denominator case - that's alright, no problem there. You certainly cannot do everything at once. 

I am writing these comments because mainly to try to make you (and your colleagues) aware of use cases that we "out in the industrial real world" have and you perhaps weren't thinking about (or didn't focus on for business reasons) when taking the decisions on how to implement certain things. Please don't understand it as a demand - "YOU MUST DO THIS!" but more a friendly reminder that some assumption may not be right. Those are not hypothetical scenarios - each has a business case behind it where I have hit a wall because of some OSVR limitation. 

E.g. re /me/head issue - apart from the Kinect example (btw, that is where I have found those hardwired VRPN buffer limits some time ago) we are discussing how to implement support for general screen configuration being provided by an OSVR server. There are different "stations" in a lab with various setups - e.g. a large screen or a "mini-cave" with 3 walls or a station with an HMD. The idea is to have a server tell the client app what they have to connect to in order to get the tracking data and how to set up their display for output when a student connects their laptop in the workspace. 

That is certainly not a use case supported out of the box, even though there is nothing really preventing the developer to have the server send the required info via the tree. However, if there are hardwired requirements around the /me/head and /display, they will likely cause server issues because that environment is multi-user and not HMD ...

When it comes to these things I prefer the KISS principle - it is often better to not implement a feature or to make it "dumb" and "manual" (=aka require the app to provide a piece of required data) than to try to be too smart and cause the developer to tear their hair out. Of course, that doesn't mean that one cannot have good default settings or a set of utilities (functions/libs/tools ...) to make the common cases simple to use.

**Comment by JeroMiya** [comment]
@janoc Thank you for the feedback. I suggested the virtual /me/head with the caveat "if it makes sense to do so" because I wasn't familiar with your requirements. It was only a suggestion. In your case I would imagine the "figure out which user is which based on on-screen directions to stand in a spot" would be functionality layered on top of OSVR rather than in the PluginKit. While it supports analysis plugins, I would generally recommend against providing functionality that is too high-level or application specific in the plugin layer.

I haven't read the discussion on specifying a port/url for the client context. However, as a temporary workaround you can configure your local server to add a reference to your other server via an external device. Here is a sample configuration that does this, along with setting up aliases to the other device:

```
{
  "externalDevices": {
    "/myExternalDevice": {
      "deviceName": "Tracker0",
      "server": "localhost:3884",
      "descriptor": { /* can also provide a file path here, or this is the minimal */
        "interfaces": {
          "tracker": {},
          "button": {},
          "analog": {}
        }
      }
    }
  },
  "aliases": {
    "/me/head": {
      "rotate": {
        "axis": "-x",
        "degrees": 180
      },
      "child": {
        "changeBasis": {
          "x": "x",
          "y": "-z",
          "z": "-y"
        },
        "child": "/myExternalDevice/tracker/0"
      }
    },
    "/me/hands/right": {
      "rotate": {
        "axis": "-x",
        "degrees": 180
      },
      "child": {
        "changeBasis": {
          "x": "x",
          "y": "-z",
          "z": "-y"
        },
        "child": "/myExternalDevice/tracker/1"
      }
    },
    "/controller/trigger": "/myExternalDevice/analog/0",
    "/controller/a": "/myExternalDevice/button/0"
  }
}
```

Note that this will cause the client to connect directly to the external server - it's not a pass-through or mirror of the other server, so there is no latency added by doing it this way.

**Comment by janoc** [comment]
@JeroMiya 
>would be functionality layered on top of OSVR rather than in the PluginKit. While it supports analysis plugins, I would generally recommend against providing functionality that is too high-level or application specific in the plugin layer.

Yes, exactly - that is the sort of functionality that certainly doesn't belong in the middleware. My point with it was to illustrate that a plugin trying to make assumptions about which user is the "main" one to assign the /me/head alias to would inevitably have problems, as it is unlikely that an application using a higher level strategy/technique would be able to use that. 

Re configuring the external server - yes, I am aware of this option and we are using it for legacy things that don't provide their native OSVR servers (but provide VRPN). However, that doesn't solve the issue when deploying to GearVR - there it is fairly difficult to start a sub-process from inside a Unity (or even a Java) app, one would have to rely on some native code hacks. Moreover, there will likely be issues with permissions and such on a non-rooted device.

I am aware of the new option to host the OSVR server in-process with the client that was added recently, but that is not something that is integrated with the Unity support so far. 

In any case, that solution just feels wrong, because the hypothetical server running on the smartphone would be only used to send the client out to the real server hosting the actual device, there is little else that could be realistically served by an OSVR server running on a phone that doesn't have a better "native" (as in Unity/Java) API already. Why to spend the precious memory, CPU and battery life on something like that? There could be other Android deployments where this could make more sense (a smart TV with connected USB hw?) but GearVR is not really that case.

**Comment by JeroMiya** [comment]
You might want to use the in-process "server" thread locally on android to access other android plugins, like the android's camera or a gamepad connected via bluetooth, etc... and to do so without having an Android-specific code-path to access those interfaces (that's pretty messy in game engines like Unity). The server thread doesn't add a significant CPU overhead over what you'd be doing yourself with the native APIs - if not, then that's probably a performance bug we'd need to address.

If you're only using it to configure an external device, you can set sleep to 1ms or more, don't deploy any auto-load plugins, and the server thread's CPU usage will drop to 0%.

**Comment by janoc** [comment]
Yes, that was sort of the idea with the in-process server thread. However, as the current Unity integration doesn't handle that, afaik. So I would have to resort to a native code hack and then to re-integrate it with Unity. 

And re other devices - yeah, camera is an enormous pain, but I wonder how did you manage to implement Android Camera 2 API which is Java-only? (the original native API is deprecated and doesn't work on many devices). BT gamepad - yes, but that is easier to handle by the built-in Unity code than going through native lib on Android ...

**Comment by russell-taylor** [comment]
@janoc I very much appreciate your feedback, it is critical to the usefulness of OSVR or any other library that the users point out limitations and improvements.  I hope that my note doesn't dampen your feedback.  We really appreciate your engagement, especially in the pre-release development while we work to get things more general.

**Comment by janoc** [comment]
@russell-taylor Thanks and you are welcome. And no worries about "dampening feedback" - I am a thick-skinned whiner :-p
